crates/autopilot/src/boundary/events/cow_amm_product_factory.rs

```diff
@@ -0,0 +1,106 @@
+use {
+    contracts::CowAmmConstantProductFactory,
+    ethcontract::{common::DeploymentInformation, dyns::DynWeb3},
+    ethrpc::current_block::RangeInclusive,
+    primitive_types::H160,
+    shared::{event_handling::EventStoring, impl_event_retrieving},
+    std::{
+        collections::{BTreeMap, HashSet},
+        sync::Arc,
+    },
+    tokio::sync::RwLock,
+};
+
+impl_event_retrieving! {
+    pub CowAmmConstantProductFactoryContract for contracts::cow_amm_constant_product_factory
+}
+
+/// CoW AMM indexer which stores events in-memory.
+#[derive(Clone)]
+pub struct Indexer {
+    /// Registry with the format: (block number, event)
+    registry: Arc<
+        RwLock<BTreeMap<u64, contracts::cow_amm_constant_product_factory::event_data::Deployed>>,
```

Comment:
Is it expected to grow infinitely? How the registry is expected to be used? Currently I see that it is accessed using `get_cow_amm_addresses` and `last_event_block` only. Does it make sense to store all the events? Can it be replaced by storing the latest block only and a HashSet with CoW AMM addresses?

Reply:
> Is it expected to grow infinitely?
Not right now. For example in mainnet there is no event at all, in gnosis chain there are only 4 events. The idea is to have a first version which stores this in-memory since it is little data and it takes almost no time to fetch it. But storing the events in the database is something to consider as a follow up PR. This won't be the only PR for the CoW AMM indexing.

This is from the task description:
> As the first step I would simply refetch all such events since contract deployment on every restart (to get the ball rolling). If this takes too much time we can consider persisting these events in our DB like we do with other events.